lib/ansible/executor/play_iterator.py

```diff
@@ -478,6 +528,19 @@ def is_failed(self, host):
         s = self.get_host_state(host)
         return self._check_failed_state(s)
 
+    def clear_host_errors(self, host):
+        self._clear_state_errors(self.get_state_for_host(host.name))
+
+    def _clear_state_errors(self, state):
+        state.fail_state = FailedStates.NONE
```

Comment:
In this playbook, only the non-failed host executes the task after clear_host_errors:
```yaml
---
- hosts: localhost
  gather_facts: no
  tasks:
    - add_host: name={{ item }} ansible_connection=local
      loop: [h1, h2]

- hosts: h1,h2
  gather_facts: no
  handlers:
    - name: sometimes_fail
      fail:
      when: 'inventory_hostname == "h1"'
  tasks:
    - command: echo
      notify: sometimes_fail
    - meta: flush_handlers
    - meta: clear_host_errors
    - debug: msg='should run for both hosts'
```

This seems to fix it:
```suggestion
        if state.fail_state == FailedStates.HANDLERS:
            state.run_state = state.pre_flushing_run_state
            state.pre_flushing_run_state = None

        state.fail_state = FailedStates.NONE
```

Reply:
Actually. The existing behavior is the same as failure in a normal task. Not sure if this is just unintuitive to me alone or a bug, but not something that needs to be addressed in this PR.

Reply:
devel:
```
PLAY [h1,h2] *********************************************************************************************************

TASK [command] *******************************************************************************************************
changed: [h2]
changed: [h1]

TASK [meta] **********************************************************************************************************

RUNNING HANDLER [sometimes_fail] *************************************************************************************
skipping: [h2]
fatal: [h1]: FAILED! => {"changed": false, "msg": "Failed as requested from task"}
```

this PR as is, better:
```
PLAY [h1,h2] *********************************************************************************************************

TASK [command] *******************************************************************************************************
changed: [h2]
changed: [h1]

TASK [meta] **********************************************************************************************************

TASK [meta] **********************************************************************************************************

RUNNING HANDLER [sometimes_fail] *************************************************************************************
fatal: [h1]: FAILED! => {"changed": false, "msg": "Failed as requested from task"}
skipping: [h2]

TASK [meta] **********************************************************************************************************

TASK [debug] *********************************************************************************************************
ok: [h2] => {
    "msg": "should run for both hosts"
}
```

this PR with proposed fix, probably correct
```
PLAY [h1,h2] *********************************************************************************************************

TASK [command] *******************************************************************************************************
changed: [h2]
changed: [h1]

TASK [meta] **********************************************************************************************************

TASK [meta] **********************************************************************************************************

RUNNING HANDLER [sometimes_fail] *************************************************************************************
fatal: [h1]: FAILED! => {"changed": false, "msg": "Failed as requested from task"}
skipping: [h2]

TASK [meta] **********************************************************************************************************

TASK [debug] *********************************************************************************************************
ok: [h2] => {
    "msg": "should run for both hosts"
}

TASK [meta] **********************************************************************************************************

TASK [debug] *********************************************************************************************************
ok: [h1] => {
    "msg": "should run for both hosts"
}
```

I think your fix is the correct thing to do.

> After thinking about it more, making a handler failure recoverable via clear_host_errors doesn't make sense since it's the last iterator state. Having a flush_handlers midplay is confusing me.

The confusing part is that technically `IteratingStates.HANDLERS` is actually not the last iterator state. If you look at the state machine in the `PlayIterator` there is nothing that would advance the state into `IteratingStates.HANDLERS`, no `state.run_state = IteratingStates.HANDLERS`. Only thing that makes it "last" is that we have implicit `flush_handlers` tasks after each section. And `flush_handlers` is the only way for `PlayIterator` to get into `IteratingStates.HANDLERS` state. And what makes this hard (and confusing) is that you can run `flush_handlers` anywhere in the play 🤯

Reply:
Although I wonder if we should think of this as clear_host_errors issue and deal with it in a separate PR?

Reply:
Fixing as a separate PR would be fine to me, totally your call. It is closer to the behavior that I'd expect than devel. If you replace the notify/flush_handlers with a failed task:
```yaml
    # - command: echo
    #   notify: sometimes_fail
    # - meta: flush_handlers
    - fail:
      when: "inventory_hostname == 'h1'"
```
the behavior on devel matches this branch (the final `- debug: msg='should run for both hosts'` still only runs for h2), and that definitely seems like a bug to me.

Then I guess rescue/always should also work for handler failures if flush_handlers is in a block (?). This example has the same issue of unrecoverable handler failure (but works on devel):

```yaml
  tasks:
  - block:
      - command: echo
        notify: sometimes_fail
      - meta: flush_handlers
    rescue:
      - debug: msg='rescue h1'
  - debug: msg='should run for both hosts'
```

On devel:
```
PLAY [localhost] ************************************************************************************************************************************************************************

TASK [add_host] *************************************************************************************************************************************************************************
changed: [localhost] => (item=h1)
changed: [localhost] => (item=h2)

PLAY [h1,h2] ****************************************************************************************************************************************************************************

TASK [command] **************************************************************************************************************************************************************************
changed: [h1]
changed: [h2]

TASK [meta] *****************************************************************************************************************************************************************************

RUNNING HANDLER [sometimes_fail] ********************************************************************************************************************************************************
fatal: [h1]: FAILED! => {"changed": false, "msg": "Failed as requested from task"}
skipping: [h2]

TASK [debug] ****************************************************************************************************************************************************************************
ok: [h1] => {
    "msg": "Hello world!"
}

TASK [debug] ****************************************************************************************************************************************************************************
ok: [h1] => {
    "msg": "should run for both hosts"
}
ok: [h2] => {
    "msg": "should run for both hosts"
}

PLAY RECAP ******************************************************************************************************************************************************************************
h1                         : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=1    ignored=0   
h2                         : ok=2    changed=1    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

This branch:
```
PLAY [localhost] ************************************************************************************************************************************************************************

TASK [add_host] *************************************************************************************************************************************************************************
changed: [localhost] => (item=h1)
changed: [localhost] => (item=h2)

PLAY [h1,h2] ****************************************************************************************************************************************************************************

TASK [command] **************************************************************************************************************************************************************************
changed: [h2]
changed: [h1]

TASK [meta] *****************************************************************************************************************************************************************************

TASK [meta] *****************************************************************************************************************************************************************************

RUNNING HANDLER [sometimes_fail] ********************************************************************************************************************************************************
fatal: [h1]: FAILED! => {"changed": false, "msg": "Failed as requested from task"}
skipping: [h2]

TASK [debug] ****************************************************************************************************************************************************************************
ok: [h2] => {
    "msg": "should run for both hosts"
}

PLAY RECAP ******************************************************************************************************************************************************************************
h1                         : ok=1    changed=1    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
h2                         : ok=2    changed=1    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

Reply:
Pushed https://github.com/ansible/ansible/pull/77955/commits/8fa40982a960b70eb8341e3e8ebd4146a27c6451 as a fix, probably needs a test too. We'll see if `ci_complete` passes.

Nice find!

Reply:
`ci_complete` passed on the above commit and I added a test for it.

Reply:
https://github.com/ansible/ansible/pull/77955/files#r942935468 is resolved in this PR.

Reply:
To correct my initial comment in the thread - clear_host_errors works across plays currently (just like refresh_inventory). The documentation is ambiguous about how it should work, but rather than "fixing" prematurely and painting us into a corner, we decided it should just mirror the existing behavior.